File: CLASS/svm/figs/svm-dual.py
#!/usr/bin/env python

import sys
import random
import numpy as np

def phi2h(x): #homogeneous quadratic mapping
    r = np.zeros(d*(d+1)/2)
    k = 0
    for i in range(d):
        for j in range(i,d):
            if i==j: r[k] = x[i]*x[j]
            else: r[k] = np.sqrt(2)*x[i]*x[j]
            k += 1
    return r

kernel = "linear" #default linear kernel
loss = "quadratic" #default, choose quadratic or hinge loss

fname = sys.argv[1]
C = np.float(sys.argv[2])
if len(sys.argv) > 3:
    kernel = sys.argv[3]
if len(sys.argv) > 4:
    loss = sys.argv[4]

print "params", fname, C, kernel, loss

D = np.loadtxt(fname)
(n, d) = np.shape(D) #get input dimensions

X = D[:,0:2]
X = np.append(X, np.ones((n,1)), 1) #add extra column of ones
Y = D[:,2] #classes

if kernel == "linear":
    K = np.dot(X, X.T)
    if loss == "quadratic":
        delta = np.diag(np.tile(1.0/(2*C), n))
        K = K+delta
elif kernel == "quadratic":
    K = np.dot(X, X.T)
    K = K*K #square each element
    if loss == "quadratic":
        delta = np.diag(np.tile(1.0/(2*C), n))
        K = K+delta

eta = 1/np.diag(K)
alpha = np.zeros(n)

#start iterative stochastic gradient ascent
eps = 0.0001
t = 0
err = 1
while err > eps:
    alpha_prev = alpha.copy()
    idx = range(n)
    #random.shuffle(idx)
    for k in idx:
        ay = alpha*Y
        ayK = np.dot(ay, K[:,k])
        alpha[k] += eta[k] * (1 - Y[k]*ayK)
        if alpha[k] < 0: alpha[k] = 0
        elif alpha[k] > C: alpha[k] = C
        
    t += 1
    err = np.linalg.norm(alpha-alpha_prev)
    print t, err

print "The support vectors are:"
for (i, ai) in enumerate(alpha):
    if ai > 0: print i, ai
print "number of support vectors:", len(alpha[alpha > 0])

if kernel == "linear":
    ay = alpha*Y
    w = np.sum(X.T*ay, axis=1)
    print "hyperplane:", w
else: #quadratic homogeneous
    Z = np.array([phi2h(x) for x in X]) #row-wise mapping to phi
    ay = alpha*Y
    w = np.sum(Z.T*ay, axis=1)
    print "w:", w
    z = phi2h(np.ones(d))
    print "h:", w*z

File: CLASS/svm/figs/svm-dual.R
#!/usr/bin/env Rscript
# gradient descent learning for SVMs
library(methods)
compute_rate <- function(maxNorm){
    R <- c(rep(0,n))
	dot <- diag(K)
	#print(dot)
	R <- 1/dot
	#print(R)
	maxNorm <<- max(dot)
	maxNorm <<- sqrt(maxNorm)

    return(R)
}

normsq <- function(x){
        return (sum(x*x))
}

get_dot_sum <- function(idx){
	AY <- A*Y
	ssum <- sum(AY*K[,idx])
	return (ssum)
}

projectval <- function(a){
	if (a < 0) return (0)
	else if (a > C) return (C)
	else return (a)
}

#homogeneous quadratic phi mapping
phi2h <- function(x){
	#since points are in 1 higher dimension, the individual
	#dimenson linear components are obtained by dot-prod with the
	#last dimension, which is 1
	r = c(rep(0,d*(d+1)/2));
	k=1;
	for (i in 1:d){
		for (j in i:d){
			if (i == j){
					r[k] = x[i]*x[j];
			}
			else{
					r[k] = sqrt(2)*x[i]*x[j];
			}
			k = k+1;
		}
	}
	return (r)
}

#########MAIN###########
args <- commandArgs(TRUE)
print(args)
trainfile = args[1]
testfile = args[2]
C <- as.real(args[3]) #regularization constant

kernel = 'linear'
if (length(args) > 3){
	kernel = args[4]
}

D <- as.matrix(read.table(trainfile))
n <- nrow(D)
d <- ncol(D)

X <- D[,1:(d-1)]
Y <- D[,d]
ones <- c(rep(1,nrow(D)))
X <- as.matrix(cbind(X,ones))

if (kernel == 'linear'){
	K <- X %*% t(X);
} else if (kernel == 'quadratic'){
	K <- X %*% t(X);
	#K = 1+K;
	K <- K*K;
}

A <- c(rep(0,n))
maxNorm <- 0
R <- compute_rate(maxNorm)

#start iterative stochastic gradient ascent
iters = 0
eps = 0.00001
repeat{
	#idxs = sample(n)
	Aold = A
	for (k in 1:n){
		#i = idxs[k]
		i = k
		g <- 1-Y[i]*get_dot_sum(i)
		A[i] <- A[i] + R[i]*g
		A[i] <- projectval(A[i])
    }
    iters <- iters + 1
	err = sqrt(normsq(A-Aold))
	show(c("err", err))
    if (err < eps ){
        break
	}
}

cat(c("\n", "Number of Iterations", iters, "\n"))
sv <- sapply(A, function(x) return (ifelse (x>0, 1, 0)))
nsv <- sum(sv)
cat("\nNumber of support vectors", nsv, "\n")
cat("\nSupport Vectors\n")
print(which(A>0))
cat("\nAlpha Values\n")
print(A[which(A>0)])

#test the model
Test <- as.matrix(read.table(testfile))
ntest <- nrow(Test)
Xt <- Test[,1:(d-1)]
Yt <- Test[,d]
ones <- c(rep(1,ntest))
Xt <- as.matrix(cbind(Xt,ones))
Kt <- X %*% t(Xt)

if (kernel == 'linear'){
	Kt <- X %*% t(Xt);
} else if (kernel == 'quadratic'){
	Kt <- X %*% t(Xt);
	#Kt = 1+Kt;
	Kt <- Kt*Kt;
}

Yh <- t(Kt)%*%(A*Y)

preds <- sapply(Yh, function(x) return (ifelse(x>0, 1, -1)))
pp <- preds*Yt
errors <- which(pp < 0)
#print(errors)
cat("\nError Rate\n")
print(length(errors)/ntest)

if (kernel == 'linear'){
	w <- t(X) %*% (A*Y) #compute the weight vector
} else if (kernel == 'quadratic'){
    phiX = apply(X, 1, phi2h)
    w = phiX %*% (A*Y)
}
cat("\nWeight Vector\n")
print(c(w))

#if (kernel == 'linear'){
	#ind = A[which(A>0 & A < C)]
	#wxy <- (X[ind,]*Y[ind]) %*% w
	#print(wxy)
	#margin = min(wxy)
	#print(margin)
	#w = w*(1/margin)
	#print(w)
#}

